File: lecture1/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

bool isEven(int n);
int len(long n);
bool isValid(long n);
int digitSum(int n);
void checkFlagCard(long creditCardNumber);
int position(long n, int digitPosition);

int main(void) 
{

    long creditNumber;
    do
    {
        creditNumber = get_long("Number: ");
    }
    while (creditNumber < 1);

    int i;

    i = isValid(creditNumber);

    if (i  == false || len(creditNumber) < 11)
    {
        printf("INVALID\n");
    }
    else 
    {
        checkFlagCard(creditNumber);
    } 
}

bool isEven(int n)
{
    if (n % 2 == 0)
    {
        return true;
    }
    else 
    {
        return false;
    }
}

int len(long n)
{
    int lenght = 0;
    do
    {
        lenght++;
        n /= 10;
    }
    while (n);
    return lenght;
}



int digitSum(int n)
{
    int result;
    result = (n * 2) % 10;
    result += (n * 2) / 10;
    return result;
}

void checkFlagCard(long creditCardNumber)
{
    int position1 = position(creditCardNumber, 1);
    int position2 = position(creditCardNumber, 2);

    if (position1 == 4)
    {
        printf("VISA\n");
    }
    else if ((position1 == 3) && (position2 == 4 || position2 == 7))
    {
        printf("AMEX\n");
    }
    else if ((position1 == 5) && (position2 > 0 && position2 < 6))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

int position(long n, int digitPosition)
{
    int lenght = len(n);
    int result;
    while (lenght > 0)
    {
        if (lenght == digitPosition)
        {
            result = n % 10;
        }

        n /= 10;
        lenght--;
    }
    return result;
}

bool isValid(long creditCardNumber)
{
    int lastNumber = creditCardNumber % 10;
    int evenSum;
    int oddSum;
    int counter = 1;
    int lenght = len(creditCardNumber);

    // while works with the lenght of credit card number
    while (lenght > 0)
    {
        // if the counter is even part of luhn's algorithm 
        // are applied (for even numbers)
        if (isEven(counter) == true)
        {
            // if number * 2 is higher than 9 digits of number
            // are addeds according luhn's algorithm
            if (lastNumber * 2 > 9)
            {
                evenSum += digitSum(lastNumber);
            }

            else
            {
                evenSum += (lastNumber * 2);
            }
        }

        //if the counter is odd, the other part of luhn's algorithm are applied
        else
        {
            oddSum += lastNumber;
        }

        // remove the last number of credit card
        creditCardNumber /= 10;

        // pick the new last number of credit card
        lastNumber = creditCardNumber % 10;

        lenght--;
        counter++;
    }

    //check the total modulo 10 is congruent to 0
    int total = 1;
    total = (evenSum + oddSum) % 10;
    if (total == 0)
    {
        return true;
    }
    if (total > 0)
    {
        return false;
    }
    else
    {
        return false;
    }
}

File: lecture6/pset6/sentimental-credit/credit.py
from re import match
from cs50 import get_int


def main():
    number = get_credit_number()
    if is_valid(number):
        check_flag(number)
    else:
        print("INVALID")

    # check_flag(number)

    test_valid(378282246310005)
    test_valid(371449635398431)
    test_valid(5555555555554444)
    test_valid(5105105105105100)
    test_valid(4111111111111111)
    test_valid(4012888888881881)
    test_valid(1234567890)
    print()
    test_is_even(1234567890, "even")
    test_is_even(2, "even")
    test_is_even(3, "odd")
    test_is_even(13, "odd")
    test_is_even(5, "odd")
    print()
    test_digit_sum(6, 3)
    test_digit_sum(8, 7)
    test_digit_sum(7, 5)
    test_digit_sum(5, 1)


def test_valid(number):
    if is_valid(number):
        print("OK")
    else:
        print("test_valid ERROR")


def test_is_even(number, expected):
    if is_even(number):
        print(f"even : {expected}")
    else:
        print(f"odd : {expected}")


def test_digit_sum(number, expected):
    result = digit_sum(number)
    if result == expected:
        print(f"{result} : {expected}")
    else:
        print(f"ERROR {result} : {expected}")


def get_credit_number():
    while True:
        change = get_int("Number: ")
        if change >= 0:
            break
    return change


def is_even(number):
    if (number % 2) == 0:
        return True
    else:
        return False


def is_valid(card_number):

    last_number = card_number % 10
    lenght = len(str(card_number))
    counter = 1
    even_sum = 0
    odd_sum = 0

    while(lenght > 0):
        if is_even(counter):
            even_sum += calculate(last_number)
        else:
            odd_sum += last_number

        card_number //= 10
        last_number = card_number % 10
        lenght -= 1
        counter += 1

    if (even_sum + odd_sum) % 10 == 0:
        return True
    else:
        return False


def digit_sum(number):
    number *= 2
    return (number // 10 % 10) + (number % 10)


def calculate(number):
    last_number = number % 10

    if (last_number * 2) > 9:
        return digit_sum(last_number)
    else:
        return last_number * 2


def check_flag(card_number):
    card_number = str(card_number)
    if match(r'^3[47]\d{13}$', card_number):
        print("AMEX")
    elif match(r'^4[0-9]{12}(?:[0-9]{3})?$', card_number):
        print("VISA")
    elif match(r'^5[1-5][0-9]{14}$', card_number):
        print("MASTER")
    else:
        print("INVALID")


main()
